pkg/verification/service/verification_service.go

```diff
@@ -0,0 +1,265 @@
+package service
+
+import (
+	"crypto/md5"
+	"crypto/rand"
+	"encoding/hex"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/kevinburke/twilio-go"
+
+	"github.com/codeready-toolchain/registration-service/pkg/application/service"
+	"github.com/codeready-toolchain/registration-service/pkg/application/service/base"
+	servicecontext "github.com/codeready-toolchain/registration-service/pkg/application/service/context"
+
+	"github.com/codeready-toolchain/registration-service/pkg/errors"
+
+	"github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"
+	"github.com/codeready-toolchain/registration-service/pkg/log"
+	"github.com/gin-gonic/gin"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+)
+
+const (
+	codeCharset = "0123456789"
+	codeLength  = 6
+
+	TimestampLayout = "2006-01-02T15:04:05.000Z07:00"
+)
+
+// ServiceConfiguration represents the config used for the verification service.
+type ServiceConfiguration interface {
+	GetTwilioAccountSID() string
+	GetTwilioAuthToken() string
+	GetTwilioFromNumber() string
+	GetVerificationMessageTemplate() string
+	GetVerificationAttemptsAllowed() int
+	GetVerificationDailyLimit() int
+	GetVerificationCodeExpiresInMin() int
+}
+
+// ServiceImpl represents the implementation of the verification service.
+type ServiceImpl struct {
+	base.BaseService
+	config     ServiceConfiguration
+	HttpClient *http.Client
+}
+
+type VerificationServiceOption func(svc *ServiceImpl)
+
+// NewVerificationService creates a service object for performing user verification
+func NewVerificationService(context servicecontext.ServiceContext, cfg ServiceConfiguration, opts ...VerificationServiceOption) service.VerificationService {
+	s := &ServiceImpl{
+		BaseService: base.NewBaseService(context),
+		config:      cfg,
+	}
+
+	for _, opt := range opts {
+		opt(s)
+	}
+	return s
+}
+
+// InitVerification sends a verification message to the specified user.  If successful,
+// the user will receive a verification SMS.
+func (s *ServiceImpl) InitVerification(ctx *gin.Context, userID string, e164PhoneNumber string) (initError error) {
+	signup, err := s.Services().SignupService().GetUserSignup(userID)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Error(ctx, err, "usersignup not found")
+			return errors.NewNotFoundError(err, "usersignup not found")
+		}
+		log.Error(ctx, err, "error retrieving usersignup")
+		return errors.NewInternalError(err, fmt.Sprintf("error retrieving usersignup: %s", userID))
+	}
+
+	// check that verification is required before proceeding
+	if signup.Spec.VerificationRequired == false {
+		log.Errorf(ctx, errors.NewForbiddenError("forbidden request", "verification code will not be sent"), "phone verification not required for usersignup: %s", userID)
```

Comment:
Why logging it as Forbidden? Why not as a bad request which is actually returned to the client?

Reply:
I'm not actually sure.. this code was taken from here: 

https://github.com/codeready-toolchain/registration-service/blob/master/pkg/controller/signup.go#L73

I'll change it to BadRequest.

Reply:
Actually, we can just log this here.  No need to log an actual error.